arch/arm/core/fault.c

```diff
@@ -273,6 +254,49 @@ static u32_t _MpuFault(NANO_ESF *esf, int fromHardFault)
 	}
 #endif /* !defined(CONFIG_ARMV7_M_ARMV8_M_FP) */
 
+#if defined(CONFIG_MPU_STACK_GUARD) || defined(CONFIG_USERSPACE)
+	/* When stack protection is enabled, we need to assess
+	 * if the memory violation error is a stack corruption.
+	 *
+	 * MemManage Faults are always banked between security
+	 * states. Therefore, we can safely assume the fault
+	 * originated from the same security state.
+	 *
+	 * As we only assess thread stack corruption, we only
+	 * process the error further if the stack frame is on
+	 * PSP. For always banked MemManage Fault, this is
+	 * equivalent to inspecting the RETTOBASE flag.
+	 *
+	 * Finally, we only process the error further, if this
+	 * is, additionally, a Stacking error.
+	 */
+	if ((SCB->ICSR & SCB_ICSR_RETTOBASE_Msk) &&
+		(SCB->CFSR & SCB_CFSR_MSTKERR_Msk)) {
+		u32_t min_stack_ptr =
+			evaluate_current_thread_stack_corruption(
+				mmfar, ((u32_t) &esf[0]));
+		if (min_stack_ptr) {
+			/* When MemManage Stacking Error has occurred, the
+			 * the stack context frame might be corrupted but
+			 * the stack pointer has actually moved. We might
+			 * face problems with un-stacking the frame, upon
+			 * exception return, if we do not have sufficient
+			 * permissions to read the corrupted stack frame.
+			 * Therefore, we move the stack manually to the
+			 * minimum allowed position for the stack pointer.
+			 */
+			__set_PSP(min_stack_ptr);
```

Comment:
@agross-linaro could you please try to test the patch on Musca_a, with and without the PSP re-adjustment in this line?